lib/datasource/github-releases/index.ts

```diff
@@ -66,11 +74,183 @@ export async function getReleases({
     })
   );
   const cacheMinutes = 10;
-  await packageCache.set(
-    cacheNamespace,
-    getCacheKey(registryUrl, repo),
-    dependency,
-    cacheMinutes
-  );
+  await packageCache.set(cacheNamespace, cacheKey, dependency, cacheMinutes);
   return dependency;
 }
+
+type DigestAsset = {
+  assetName: string;
+  currentVersion: string;
+  currentDigest: string;
+  digestedFileName?: string;
+};
+
+async function findDigestFile(
+  release: GithubRelease,
+  digest: string
+): Promise<DigestAsset | null> {
+  const smallAssets = release.assets.filter(
+    (a: GithubReleaseAsset) => a.size < 5 * 1024
+  );
+  for (const asset of smallAssets) {
+    const res = await http.get(asset.browser_download_url);
+    for (const line of res.body.split('\n')) {
+      const [lineDigest, lineFn] = line.split(/\s+/, 2);
+      if (lineDigest === digest) {
+        return {
+          assetName: asset.name,
+          digestedFileName: lineFn,
+          currentVersion: release.tag_name,
+          currentDigest: lineDigest,
+        };
+      }
+    }
+  }
+  return null;
+}
+
+function inferHashAlg(digest: string): string {
+  switch (digest.length) {
+    case 64:
+      return 'sha256';
+    default:
+    case 96:
+      return 'sha512';
+  }
+}
+
+async function findAssetWithDigest(
+  release: GithubRelease,
+  digest: string
+): Promise<DigestAsset | null> {
+  const algorithm = inferHashAlg(digest);
+  const assetsBySize = release.assets.sort(
+    (a: GithubReleaseAsset, b: GithubReleaseAsset) => {
+      if (a.size < b.size) {
+        return -1;
+      }
+      if (a.size > b.size) {
+        return 1;
+      }
+      return 0;
+    }
+  );
+
+  for (const asset of assetsBySize) {
+    const res = http.stream(asset.browser_download_url);
+    const assetDigest = await hasha.fromStream(res, { algorithm });
+    if (assetDigest === digest) {
+      return {
+        assetName: asset.name,
+        currentVersion: release.tag_name,
+        currentDigest: assetDigest,
+      };
+    }
+  }
```

Comment:
This should be agressively cached. as this can be huge downloads and expensive hash computation.